### PR TITLE
Develop poca cc num agents

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

Adding the normalized number of agents as input to the POCA centralized critic. 
Making experiments on the spawning environment. Will post results soon.
Number of agents is fed as input as a single float between -1 and 1 where -1 is a single agent and 1 is the maximum number of agents seen so far.

Screen Shot 2021-07-12 at 08 55 19

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/main/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/main/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/main/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/poca/optimizer_torch.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def __init__(
             else:
                 encoding_size = network_settings.hidden_units
 
-            self.value_heads = ValueHeads(stream_names, encoding_size, 1)
+            self.value_heads = ValueHeads(stream_names, encoding_size + 1, 1)
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/poca/optimizer_torch.py:63]: maybe a comment "+1 for num agents input"

File: ml-agents/mlagents/trainers/poca/optimizer_torch.py [hunk 2]
```diff
@@ -95,6 +95,12 @@ def baseline(
                 memories=memories,
                 sequence_length=sequence_length,
             )
+
+            num_agents = self.network_body.agent_count(
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/poca/optimizer_torch.py:99]: I wonder if we should just add this to the buffer and calculate it when processing trajectories?

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 3]
```diff
@@ -354,6 +357,27 @@ def _copy_and_remove_nans_from_obs(
             obs_with_no_nans.append(no_nan_obs)
         return obs_with_no_nans
 
+    def agent_count(
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/torch/networks.py:360]: I think this is a pretty complicated function but there are loops in the agent processor that have this information e.g. https://github.com/Unity-Technologies/ml-agents/blob/main/ml-agents/mlagents/trainers/agent_processor.py#L246. I think we can just add it to the buffer right there as an integer and then normalize in the trainer/network body

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/networks.py:360]: cc @ervteng

Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py:360]: I'd be OK with any approach that doesn't call `get_masks_from_nans` again, as it's not the cheapest of functions for large inputs. I think we could move the call for this method into the `forward` method of MultiAgentNetworkBody perhaps, there we can re-use the mask that is already computed

Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py:360]: Adding it to the buffer seems redundant IMO, but it's not the worst

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/networks.py]: I simplified the logic a bunch. I reuse `get_masks_form_nans` now.

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 4]
```diff
@@ -404,12 +407,22 @@ def forward(
...
         encoded_state = self.self_attn(encoded_entity, self_attn_masks)
 
+        flipped_masks = 1 - torch.cat(self_attn_masks, dim=1)
+        num_agents = torch.sum(flipped_masks, dim=1, keepdim=True)
+        if torch.max(num_agents).item() > self._current_max_agents:
+            self._current_max_agents = torch.nn.Parameter(
+                torch.as_tensor(torch.max(num_agents).item()), requires_grad=False
+            )
+
+        num_agents = num_agents * 2.0 / self._current_max_agents - 1
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/torch/networks.py:417]: Will probably want a comment on this line of math for posterity

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 5]
```diff
@@ -404,12 +407,22 @@ def forward(
         encoded_entity = torch.cat(self_attn_inputs, dim=1)
         encoded_state = self.self_attn(encoded_entity, self_attn_masks)
 
+        flipped_masks = 1 - torch.cat(self_attn_masks, dim=1)
+        num_agents = torch.sum(flipped_masks, dim=1, keepdim=True)
+        if torch.max(num_agents).item() > self._current_max_agents:
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/torch/networks.py:412]: I'm a little concerned with this O(N) operation (`torch.max`) but if it doesn't really affect runtime then it's fine

**Review by ervteng — APPROVED** [review]
Left some small comments

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 6]
```diff
@@ -404,12 +407,22 @@ def forward(
...
+            self._current_max_agents = torch.nn.Parameter(
+                torch.as_tensor(torch.max(num_agents).item()), requires_grad=False
+            )
+
+        num_agents = num_agents * 2.0 / self._current_max_agents - 1
+
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/networks.py:418]:
```suggestion
            )
        
        # num_agents will be -1 for a single agent and +1 when the current maximum is reached
        num_agents = num_agents * 2.0 / self._current_max_agents - 1

```
